### PR TITLE
feat: migrate codebox component

### PR DESCRIPTION
**Pull request by araujogui**
Issue #5242
Migrate Codebox component

### CODE REVIEW

**Comment by ovflowd** [comment]
@araujogui please run `npx turbo lint` your branch has linting issues 🤔

File: .storybook/preview.tsx [hunk 1]
```diff
@@ -23,6 +23,7 @@ const preview: Preview = {
         basePath: '',
       },
     },
+    backgrounds: { disable: true },
```
Comment (ovflowd) [inline .storybook/preview.tsx:26]: Just a note here: This change was done because we have our own Theming System. We might want to integrate Storybook to support our `next/themes` stuff

File: .browserslistrc [hunk 2]
```diff
@@ -1,14 +1,10 @@
...
 [production]
 
 cover 95%
 not dead
 
 [development]
 
-defaults and supports es6-module
-maintained node versions
+defaults
```
Comment (ovflowd) [inline .browserslistrc]: This fixes `storybook` local development mode. @araujogui could you make another PR just with the changes to this file? That'd be great!

File: components/Article/Codebox/InlineCode.tsx [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import styles from './index.module.scss';
+import type { PropsWithChildren } from 'react';
+
+const InlineCode = ({ children }: PropsWithChildren) => (
```
Comment (ovflowd) [inline components/Article/Codebox/InlineCode.tsx]: Could you change this to `const InlineCode: React.FC<PropsWithChildren> = ({ children }) =>`

File: components/Article/Codebox/__tests__/index.test.tsx [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    expect(replaceLanguages('language-cjs')).toBe('language-js');
+    expect(replaceLanguages('language-javascript')).toBe('language-js');
+    expect(replaceLanguages('language-console')).toBe('language-bash');
+    expect(replaceLanguages('language-shell')).toBe('language-bash');
+  });
+});
+
+describe('Codebox component (one lang)', (): void => {
+  const code = 'const a = 1;';
+  const textToCopy = sanitize(highlight(code, languages.js, 'js'));
```
Comment (ovflowd) [inline components/Article/Codebox/__tests__/index.test.tsx]: I think this should be removed, we're double-encoding the content

File: components/Article/Codebox/__tests__/index.test.tsx [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    render(
+      <IntlProvider locale="en" onError={() => {}}>
+        <Codebox>
+          <code className="language-cjs|language-mjs">{code}</code>
+        </Codebox>
+      </IntlProvider>
+    );
+
+    const buttonElement = screen.getByText('cjs');
+    userEvent.click(buttonElement);
```
Comment (ovflowd) [inline components/Article/Codebox/__tests__/index.test.tsx]: This should be async with `const user = userEvent.setup()`

File: components/Article/Codebox/__tests__/index.test.tsx [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    render(
+      <IntlProvider locale="en" onError={() => {}}>
+        <Codebox>
+          <code className="language-async|language-sync">{code}</code>
+        </Codebox>
+      </IntlProvider>
+    );
+
+    const buttonElement = screen.getByText('async');
+    userEvent.click(buttonElement);
```
Comment (ovflowd) [inline components/Article/Codebox/__tests__/index.test.tsx]: Same here about userEvent

File: components/Article/Codebox/index.module.scss [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+  tab-size: 4;
+  text-align: left;
+  white-space: pre;
+  word-break: normal;
+  word-spacing: normal;
+  word-wrap: normal;
+}
+
+.code {
+  @extend %baseStyles;
```
Comment (ovflowd) [inline components/Article/Codebox/index.module.scss]:
```suggestion
  @extend %baseStyles;

```

File: components/Article/Codebox/index.stories.tsx [hunk 8]
```diff
@@ -0,0 +1,11 @@
+import Codebox from './index';
+
+export default { component: Codebox };
+
+const code = 'const a = 1;';
+
+export const Default = {
+  args: {
+    children: <code className="language-js">{code}</code>,
```
Comment (ovflowd) [inline components/Article/Codebox/index.stories.tsx]:
```suggestion
    children: <pre className="language-js">{code}</pre>,
```

Comment (ovflowd) [reply components/Article/Codebox/index.stories.tsx]: If it is `code` then it should be the InLineCode. Also could you add another story here for the InLine code?

File: components/Article/Codebox/index.tsx [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+  language
+    .replace(/mjs|cjs|javascript/i, 'js')
+    .replace(/console|shell/i, 'bash');
+
+const Codebox: FC<CodeBoxProps> = ({ children: { props } }) => {
+  const [parsedCode, setParsedCode] = useState('');
+  const [copied, copyText] = useCopyToClipboard();
+  const [langIndex, setLangIndex] = useState(0);
+
+  const className = props.className || 'text';
```
Comment (ovflowd) [inline components/Article/Codebox/index.tsx:27]:
```suggestion
  const className = props.className || 'text';

```

File: components/Article/Codebox/index.tsx [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+  // Language Matches in class
+  const matches = className.match(/language-(?<lang>.*)/);
+  const languageOptions = (matches?.groups?.lang || 'text').split('|');
+  const language = languageOptions[langIndex];
```
Comment (ovflowd) [inline components/Article/Codebox/index.tsx:33]:
```suggestion
  const language = languageOptions[langIndex];

```

File: hooks/useCopyToClipboard.ts [hunk 11]
```diff
@@ -0,0 +1,31 @@
+import { useEffect, useState } from 'react';
+
+const copyToClipboard = (value: string) => {
```
Comment (ovflowd) [inline hooks/useCopyToClipboard.ts]: Oh, I guess Augustin was also adding this same hook to their PR.

File: hooks/useCopyToClipboard.ts [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+  if (typeof window === 'undefined') {
```
Comment (ovflowd) [inline hooks/useCopyToClipboard.ts]:
```suggestion
  if (typeof navigator === 'undefined') {
```

Since we use `navigator` call down there

File: hooks/useCopyToClipboard.ts [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+  }
+
+  return navigator.clipboard
+    .writeText(value)
+    .then(() => true)
+    .catch(() => false);
+};
+
+// eslint-disable-next-line no-unused-vars
+export const useCopyToClipboard = (): [boolean, (text: string) => void] => {
```
Comment (ovflowd) [inline hooks/useCopyToClipboard.ts]: Could you remove this explicit return type and then remote the eslint-disable statement

```suggestion
export const useCopyToClipboard = () => {
```

File: hooks/useCopyToClipboard.ts [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+    if (!copied) {
+      return undefined;
+    }
+
+    const timerId = setTimeout(() => setCopied(false), 3000);
+
+    return () => clearTimeout(timerId);
+  }, [copied]);
+
+  return [copied, copyText];
```
Comment (ovflowd) [inline hooks/useCopyToClipboard.ts]:
```suggestion
  return [copied, copyText] as const;
```

File: hooks/useCopyToClipboard.ts [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+};
+
+// eslint-disable-next-line no-unused-vars
+export const useCopyToClipboard = (): [boolean, (text: string) => void] => {
+  const [copied, setCopied] = useState(false);
+
+  const copyText = (text: string) => copyToClipboard(text).then(setCopied);
+
+  useEffect(() => {
+    if (!copied) {
```
Comment (ovflowd) [inline hooks/useCopyToClipboard.ts]: Change the logic to:

```jsx
if (copied) {
  const timerId = setTimeout(() => setCopied(false), 3000);

  return () => clearTimeout(timerId);
}

return undefined;
```

**Review by ovflowd — COMMENTED** [review]
Really good stuff! Thank you for the contribution, @araujogui!!

Left a bunch of comments that I'd like to have them addressed before we proceed :)

File: hooks/useCopyToClipboard.ts (hunk 15 above)
Comment (araujogui) [reply hooks/useCopyToClipboard.ts]: An early return sounds cleaner to me

File: components/Article/Codebox/__tests__/index.test.tsx (hunk 5 above)
Comment (araujogui) [reply components/Article/Codebox/__tests__/index.test.tsx]: `userEvent.setup` is conflicting our `navigator` mocks
```
  ● Codebox component (multiple langs) › switch between languages

    TypeError: Cannot redefine property: clipboard
        at Function.defineProperty (<anonymous>)
```

Comment (ovflowd) [reply components/Article/Codebox/__tests__/index.test.tsx]: Then we don't need a mock if we do userEvent.setup(). And probably there's an API from testing library to test this navigator API.

File: components/Article/Codebox/__tests__/index.test.tsx [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+    fireEvent.click(buttonElement);
+
+    expect(navigatorClipboardWriteTextSpy).toHaveBeenCalledTimes(1);
+    expect(navigatorClipboardWriteTextSpy).toHaveBeenCalledWith(code);
+  });
+});
+
+describe('Codebox component (multiple langs)', (): void => {
+  const code = `const http = require('http');
+  -------
```
Comment (ovflowd) [inline components/Article/Codebox/__tests__/index.test.tsx]: This sounds wrong. Even the snapshot is wrong

```tsx
const code = `const http = require('http');
-------
import http from 'http';`;
```

That's how this should look like.

File: components/Article/Codebox/__tests__/index.test.tsx [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+
+    const buttonElement = screen.getByText('cjs');
+    userEvent.click(buttonElement);
+
+    await screen.findByText('cjs');
+
+    expect(screen.getByText('cjs')).toBeInTheDocument();
+  });
+
+  it('switch between sync and async', async () => {
```
Comment (ovflowd) [inline components/Article/Codebox/__tests__/index.test.tsx]: Also, remove this test. This is kind of garbage. Already covered by previous test.

File: components/Article/Codebox/index.tsx [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+
+  const codeArray = props.children
+    ? props.children.toString().split('--------------\n')
+    : [''];
+
+  const handleCopyCode = async (event: MouseEvent<HTMLButtonElement>) => {
```
Comment (ovflowd) [inline components/Article/Codebox/index.tsx]: This doesn't need to be `async`. There's no async operation happening here.

File: hooks/useCopyToClipboard.ts (hunk 15 above)
Comment (ovflowd) [reply hooks/useCopyToClipboard.ts]: Negated conditions are usually a bad practice if you can do the same condition without a negated operator. It's usually easier to read.

We often do negated conditions in JSX, because well, adding the `return null` after a gigantic block of JSX hinders the visibility of that statement.

File: jest.setup.mjs [hunk 19]
```diff
@@ -1 +1,5 @@
+import { TextEncoder, TextDecoder } from 'util';
 import '@testing-library/jest-dom/extend-expect';
+
+global.TextEncoder = TextEncoder;
```
Comment (ovflowd) [inline jest.setup.mjs]: I honestly don't like to simply inject Node's TextEncoder API to the `globals`. That's a hack and shouldn't be necessary.

You can simply mock `isomorphic-dompurify` and that will have the same effect.

Since there's nothing to be "purified" on our unit-test. (We don't have malicious code)

File: components/Article/Codebox/index.tsx (hunk 18 above)
Comment (araujogui) [reply components/Article/Codebox/index.tsx]: Actually, `copyText` is a async operation. Maybe an `await` here would be a better option?

Comment (ovflowd) [reply components/Article/Codebox/index.tsx]: Well that makes sense. But what will the onClick operator await for? Isn't onClick itself synchronous? (If I recall it does indeed in some certain blocking operations wait for something actually)

File: components/Article/Codebox/__tests__/index.test.tsx [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+    const buttonElement = screen.getByText('components.codeBox.copy');
+    await user.click(buttonElement);
+
+    expect(navigatorClipboardWriteTextSpy).toHaveBeenCalledTimes(1);
+    expect(navigatorClipboardWriteTextSpy).toHaveBeenCalledWith(code);
+  });
+});
+
+describe('Codebox component (multiple langs)', (): void => {
+  const code = `const http = require('http');
```
Comment (ovflowd) [inline components/Article/Codebox/__tests__/index.test.tsx:55]: Snapshot still seems wrong. Can you try `const http = "require('http');\n--------------\nimport http from 'http';`

Comment (ovflowd) [reply components/Article/Codebox/__tests__/index.test.tsx:55]: Ah I know why. It needs to be `--------------`

Comment (ovflowd) [reply components/Article/Codebox/__tests__/index.test.tsx:55]: I wonder how these tests are passing if they never then actually tested what this was supposed to test -.-

(Well they are, just not testing whatever they were supposed to test)

File: components/Article/Codebox/__tests__/index.test.tsx [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+        <Codebox>
+          <pre className="language-cjs|language-mjs">{code}</pre>
+        </Codebox>
+      </IntlProvider>
+    );
+
+    const buttonElement = screen.getByText('cjs');
+    await user.click(buttonElement);
+
+    await screen.findByText('cjs');
```
Comment (ovflowd) [inline components/Article/Codebox/__tests__/index.test.tsx]: Once you update that snippet. This should probably start to fail. As it should switch to `await screen.findByText('mjs');` and `expect(screen.getByText('mjs')).toBeInTheDocument();`

File: components/Article/Codebox/index.tsx [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+
+const Codebox: FC<CodeBoxProps> = ({ children: { props } }) => {
+  const [parsedCode, setParsedCode] = useState('');
+  const [copied, copyText] = useCopyToClipboard();
+  const [langIndex, setLangIndex] = useState(0);
+
+  const className = props.className || 'text';
+
+  // Language Matches in class
+  const matches = className.match(/language-(?<lang>.*)/);
```
Comment (ovflowd) [inline components/Article/Codebox/index.tsx]: This is ultimately broken. This should honestly be:

```tsx
const language = useMemo(() => {
  let language = 'text';

  if (props.className && props.className.length) {
    language = props.className.split('|')[langIndex];
  }

  if (language.startsWith('language-')) {
    language = language.replace('language-', '');
  }

  return language;
}, [props.className, langIndex]);
```

File: components/Article/Codebox/__tests__/index.test.tsx [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+      <IntlProvider locale="en" onError={() => {}}>
+        <Codebox>
+          <pre className="language-js|language-js">{code}</pre>
+        </Codebox>
+      </IntlProvider>
+    );
+    expect(container).toMatchSnapshot();
+  });
+
+  it('switch between languages', async () => {
```
Comment (ovflowd) [inline components/Article/Codebox/__tests__/index.test.tsx:59]: Can we also add a snapshot for before and after click?

**Review by ovflowd — APPROVED** [review]
Now we're talking! This PR not only migrates the component but also fixed many invisible bugs we didn't even notice.

Amazing job, @araujogui! Thank you for being attentive :)

**Comment by ovflowd** [comment]
Note: I feel that the current Storybook args is kinda not optimal.. If we could somehow allow people to actually write the content and map that to the story component?

E.g.:

- Both stories have a Decorator https://storybook.js.org/docs/react/writing-stories/decorators/
  - This decorator actually adds the `<pre>` tag and className
- Both stories then have args for:
  - Name of programming language as array
  - Content as array

Then for each programming language array item we join them with `|` and for each content we join them with that big `-------------`.

Could you do that? If it's not clear let me know!

File: __mocks__/isomorphic-dompurify.js [hunk 24]
```diff
@@ -0,0 +1,5 @@
+const mock = {
```
Comment (ovflowd) [inline __mocks__/isomorphic-dompurify.js]: The mock doesn't need to be done here. You can do it directly on the test. Since it's the only place we need to mock it for now :)

(I actually have the mock actually be done there, as it allows me to remember why this mock exists)

File: components/Article/Codebox/InlineCode.tsx
Comment (ovflowd) [inline components/Article/Codebox/InlineCode.tsx:1]: One last thing. Can you:

- Move this component to `components/Article/InlineCode`
- Move also the stories
- Move all the styles that are only for the `.code` class (including dark styles, light ones and etc cetera) also to the Inline code
- Add unit test just that it is rendering fine

That's pretty much my final ask :)

**Comment by ovflowd** [comment]
Also @araujogui I know it's not of the scope of this PR but it would be great if the currently selected programming language is highlighted either by a different colour or something...

image

Right now I can't really see what one is selected. You could add a new class to the button called selected and add this className if the current button index === selectedIndex

File: components/Article/Codebox/inline.stories.tsx [hunk 25]
```diff
@@ -0,0 +1,19 @@
...
+
+export default meta;
+
+type Story = StoryObj<typeof InlineCode>;
+
+const code = 'const a = 1;';
+
+export const Default: Story = {
+  args: {
+    children: <code>{code}</code>,
```
Comment (ovflowd) [inline components/Article/Codebox/inline.stories.tsx]: Also do the same with here (with the decorator) so we can actually customise the code xD

File: components/Article/Codebox/InlineCode.tsx
Comment (araujogui) [reply components/Article/Codebox/InlineCode.tsx:1]: Can I remove `.code` classes from Codebox, or should I leave it duplicated?

Comment (ovflowd) [reply components/Article/Codebox/InlineCode.tsx:1]: Remove. Not used there :)

File: components/Article/Codebox/index.module.scss [hunk 26]
```diff
@@ -0,0 +1,105 @@
...
+    .copy {
+      align-items: center;
+      cursor: pointer;
+      display: inherit;
+      font-size: var(--font-size-code);
+      height: 23px;
+      justify-content: center;
+      width: 86px;
+    }
+    .langBox {
```
Comment (ovflowd) [inline components/Article/Codebox/index.module.scss:24]:
```suggestion

    .langBox {
```

File: components/Article/InlineCode/index.module.scss [hunk 27]
```diff
@@ -0,0 +1,58 @@
+%baseStyles {
```
Comment (ovflowd) [inline components/Article/InlineCode/index.module.scss]: I think these base styles instead of duplicated can be outsourced to somewhere else. Like maybe styles/code.scss. And maybe renamed to %codeBaseStyles I dunno

File: components/Article/InlineCode/index.stories.tsx [hunk 28]
```diff
@@ -0,0 +1,30 @@
+import InlineCode from './index';
+import type { FC } from 'react';
+import type { Meta, StoryObj } from '@storybook/react';
+
+type InlineCodeArgs = {
+  code: string;
+};
+
+const meta: Meta<FC<InlineCodeArgs>> = {
+  title: 'InlineCode',
```
Comment (ovflowd) [inline components/Article/InlineCode/index.stories.tsx]: No need for title

File: components/Article/Codebox/index.stories.tsx [hunk 29]
```diff
@@ -0,0 +1,48 @@
...
+
+type CodeBoxProps = {
+  language: string[];
+  code: string[];
+};
+
+type ICodebox = FC<CodeBoxProps>;
+
+const meta: Meta<ICodebox> = {
+  title: 'Codebox',
```
Comment (ovflowd) [inline components/Article/Codebox/index.stories.tsx]: Same here, no title please. Otherwise we cannot group these components per folder

File: components/Article/InlineCode/index.stories.tsx [hunk 30]
```diff
@@ -0,0 +1,30 @@
...
+};
+
+export default meta;
+
+type Story = StoryObj<FC<InlineCodeArgs>>;
+
+const code = 'const a = 1;';
+
+export const Default: Story = {
+  args: {
```
Comment (ovflowd) [inline components/Article/InlineCode/index.stories.tsx]: args: { code } 

(Just in one line)

**Review by ovflowd — COMMENTED** [review]
Forgot to post my comments. Last review ;D

File: components/Article/InlineCode/index.stories.tsx (hunk 28 above)
Comment (araujogui) [reply components/Article/InlineCode/index.stories.tsx]: A story should contain at least a title or a component
```
Error: CSF: missing title/component /Users/araujo/Documents/Projects/nodejs.org/components/Article/InlineCode/index.stories.tsx (line 1, col 0)
```

Comment (ovflowd) [reply components/Article/InlineCode/index.stories.tsx]: You should pass the component then, but not title. We're not using title anywhere :)

File: hooks/__tests__/useCopyToClipboard.test.tsx [hunk 31]
```diff
@@ -0,0 +1,83 @@
...
+import type { AppProps } from '../../types';
+
+const i18nData = { currentLocale: { code: 'en' } } as AppProps['i18nData'];
+
+describe('useCopyToClipboard', () => {
+  const HookRenderer = ({ text }: { text: string }) => {
+    const [copied, copyText] = useCopyToClipboard();
+
+    return (
+      <LocaleProvider i18nData={i18nData}>
```
Comment (ovflowd) [inline hooks/__tests__/useCopyToClipboard.test.tsx]: Can you remove the locale provider from here? (No need to use any translation key. Inside the unit test component you can even write "copy text test"

File: components/Article/InlineCode/index.stories.tsx [hunk 32]
```diff
@@ -0,0 +1,28 @@
+import InlineCode from './index';
+import type { FC } from 'react';
+import type { Meta, StoryObj } from '@storybook/react';
+
+type InlineCodeArgs = {
```
Comment (ovflowd) [inline components/Article/InlineCode/index.stories.tsx]: This typing feels weird. Please check the CONTRIBUTING guidelines (there should be a Meta type)

File: components/Article/InlineCode/index.stories.tsx [hunk 33]
```diff
@@ -0,0 +1,28 @@
...
+  decorators: [
+    (Story, context) => (
+      <InlineCode>
+        <code>{context.args.code}</code>
+      </InlineCode>
+    ),
+  ],
+};
+
+export default meta;
```
Comment (ovflowd) [inline components/Article/InlineCode/index.stories.tsx]: This export should be below everything and the naming of the export should be consistent as from the template on the CONTRIBUTING guidelines (same for the other story)

File: hooks/__tests__/useCopyToClipboard.test.tsx (hunk 31 above)
Comment (araujogui) [reply hooks/__tests__/useCopyToClipboard.test.tsx]: It's a must
```
[React Intl] Could not find required `intl` object. <IntlProvider> needs to exist in the component ancestry.
```

Comment (ovflowd) [reply hooks/__tests__/useCopyToClipboard.test.tsx]: Yes because you're passing FormattedMessage there. You can simply remove that altogether lol

**Review by ovflowd — APPROVED** [review]
@nodejs/website can y'all also review/approve this?

**Comment by mikeesto** [comment]
Has something happened with our Vercel storybook previews here? All the links are 404s for me

**Comment by ovflowd** [comment]
> Has something happened with our Vercel storybook previews here? All the links are 404s for me

Misconfiguration on Vercel-side, but it was fixed already. I can redeploy this build.

**Comment by ovflowd** [comment]
@araujogui there are some conflicts here

**Comment by ovflowd** [comment]
@araujogui can you add a litle hover effect (darker colour) when you hover on the copy and language selection buttons?

**Comment by ovflowd** [comment]
cc @nodejs/website also for review

**Review by mikeesto — APPROVED** [review]
The hover effect looks good. This LGTM, just a couple of conflicts to walk through

**Review by shanpriyan — APPROVED** [review]
LGTM! 🚀 


Suggestion: 
Can we add a **green** `border-bottom` to the selected language
image

**Review by aymen94 — APPROVED** [review]

**Comment by ovflowd** [comment]
@araujogui updates?

**Comment by ovflowd** [comment]
@araujogui can you run the linting commands and update the test snapshots here? @Harkunwar is anything else needed here? (Like Unit Tests or etc)

**Comment by ovflowd** [comment]
@araujogui can you please run build/lint/test before committing? PR is still failing (build failing) with missing types?

File: components/Article/Codebox/__tests__/index.test.tsx [hunk 34]
```diff
@@ -0,0 +1,99 @@
...
+    expect(navigatorClipboardWriteTextSpy).toHaveBeenCalledWith(code);
+  });
+});
+
+describe('Codebox component (multiple langs)', (): void => {
+  const code = `const http = require('http');
+--------------
+import http from 'http';`;
+
+  it('renders correctly', (): void => {
```
Comment (ovflowd) [inline components/Article/Codebox/__tests__/index.test.tsx]: @araujogui this test can be removed, as Storybooks now have their own snapshot testing. You can also remove this snapshot,

Don't forget to run `npm run test:storybook:snapshot` just to double check all Storybooks have Snapshots
